### ISSUE TITLE
[<rllib>] Training on Repeated observation spaces broken (?); check in SingleAgentEpisode.add_env_reset fails (due to batching behavior?)

### ISSUE THREAD

**Issue by MatthewCWeston** [ray-project/ray#53327]
### What happened + What you expected to happen

As best I can tell, `ray.rllib.utils.spaces.repeated.Repeated` observation spaces don't get handled correctly when it comes to vector environments' batching of observations. When I tried to run `algo.step()` on an environment with a Repeated observation space, the following occurred:

In `SingleAgentEnvRunner._reset_envs`, the following code is run when an environment is reset:

```
observations, infos = self._try_env_reset()
observations = unbatch(observations)

# Set initial obs and infos in the episodes.
for env_index in range(self.num_envs):
    episodes[env_index].add_env_reset(
        observation=observations[env_index],
        infos=infos[env_index],
    )
```

In `SingleAgentEpisode,.add_env_reset`, `observations` , as set above, is checked against `self.observation_space`:

```
if self.observation_space is not None:
            assert self.observation_space.contains(observation), (
                f"`observation` {observation} does NOT fit SingleAgentEpisode's "
                f"observation_space: {self.observation_space}!"
            )
```

For `Repeated` observation spaces, this check fails, making it impossible to train a model on an environment that uses one, as far as I can tell. I looked into the issue, and I think it has something to do with the way `SyncVectorEnv` converts Repeated spaces into `Tuple(Repeated)` spaces.

I've cleaned up maximally-simplified code to replicate this issue and attached it below, as a reproduction script:

### Versions / Dependencies

For an identical environment to one that I've tested this on, running `!pip install -q "ray[rllib]"` in Google Colab should work well.

### Reproduction script

import torch
import gymnasium as gym
from gymnasium.spaces import Discrete, Box
from ray.rllib.utils.spaces.repeated import Repeated

from ray.tune.registry import register_env
from ray.rllib.algorithms.ppo.torch.default_ppo_torch_rl_module import DefaultPPOTorchRLModule

from ray.rllib.core.models.configs import ModelConfig, ActorCriticEncoderConfig, MLPHeadConfig
from ray.rllib.core.models.torch.base import TorchModel
from ray.rllib.core.models.base import Encoder, ENCODER_OUT
from ray.rllib.core.rl_module.rl_module import RLModuleSpec

class TestRepeatedEnv(gym.Env): # stub environment with Repeated observations
    def __init__(self, config=None):
        super().__init__()
        # Setting the observation space to Repeated causes the error. Setting it to Box results in no error occurring
        self.observation_space = Repeated(Box(-1,1,shape=(1,)), 5)
        #self.observation_space = Box(-1,1,shape=(5,))
        self.action_space = Discrete(2)

    def step(self, action):
        self.timer += 1
        done = self.timer == 10
        return self.observation_space.sample(), 0, done, False, {}

    def reset(self, *, seed=None, options=None):
        self.timer = 0
        return self.observation_space.sample(), {}
register_env("env", lambda cfg: TestRepeatedEnv())

class RepeatedPPORLModule(DefaultPPOTorchRLModule): # stub module that should produce valid outputs for arbitrary observations
    def setup(self):
        custom_config = RepeatedEncoderConfig()
        # The encoder is a preprocessor applied to the base observation before it is fed to the actor and the critic.
        actor_critic_encoder_config = ActorCriticEncoderConfig(
            base_encoder_config=custom_config
        )
        self.encoder = actor_critic_encoder_config.build(framework="torch")
        encoder_output_size = (100,) # meaningless; could be anything.
        # These are just feedforward neural networks for the actor and critic. We need to specifiy their dimensions because our encoder decides what shape of input they'll get.
        pi_config = MLPHeadConfig(
            input_dims=encoder_output_size,
            output_layer_dim=2,
        )
        vf_config = MLPHeadConfig(
            input_dims=encoder_output_size, output_layer_dim=1
        )
        self.pi = pi_config.build(framework="torch")
        self.vf = vf_config.build(framework="torch")
class RepeatedEncoderConfig(ModelConfig):
    def build(self, framework):
        return RepeatedEncoder(self)
class RepeatedEncoder(TorchModel, Encoder):
    def _forward(self, input_dict, **kwargs):
        N = input_dict['obs'].shape[0]
        return {ENCODER_OUT: torch.rand(N,100)} # arbitrary output
		
from ray.rllib.algorithms.ppo import PPOConfig
config = (
        PPOConfig()
        .environment(
            env="env",
            env_config={},
        )
        .framework("torch")
        .env_runners(
            num_env_runners=0,
            num_envs_per_env_runner=1,
        )
        .training(
            train_batch_size=256,
        )
        .rl_module(rl_module_spec=RLModuleSpec(module_class=RepeatedPPORLModule))
    )
algo = config.build()

algo.step() # An error will occur here.

### Issue Severity

High: It blocks me from completing my task.

**Comment by MatthewCWeston** [comment]
Wrote a pull request (#53519) containing my current workaround and an example demonstrating its use. Made it as flexible as possible in hopes of being of use to others.
